### PR TITLE
Add API to get contig from region of named chromosome

### PR DESCRIPTION
**Pull request by nspope**
Mostly addresses #1346, #670, #401 (and supersedes #402) by adding a way to get a contig from an arbitrary interval of a named chromosome.

Thanks to `msprime.RateMap.slice` this was pretty straightforward, except for the handling of coordinates when DFEs are added. I opted to use the same coordinate system as the chromosome from which the contig is extracted, so for example the following adds a DFE in the middle of the contig:
```python
species = stdpopsim.get_species("HomSap")
contig = species.get_contig(chromosome="chr1", left=1e6, right=1.1e6)
contig.add_dfe(intervals=[[1.05e6, 1.06e6]], dfe=...)
```
IMO this is cleaner than requiring the user to manually shift the DFE coordinates (and is less confusing with DFE bed files, annotations, etc).

Previously, if an added DFE had an interval that fell outside the contig, a rather ambiguous SLiM error would get raised. Now, the added DFE intervals are silently clipped to the contig boundaries. If all the intervals fall outside of these boundaries, a warning is raised and a DFE with empty intervals is added.

The same goes for masks (previously an error would get thrown from tskit if a mask interval overlapped a contig boundary): these are now clipped, with a warning if all masked intervals fall outside of the contig boundaries.

### CODE REVIEW

**Comment by nspope** [comment]
Ah, there's a slight issue. If the recombination map is longer than the chromosome, the current behavior of stdpopsim is to use the recombination map length for simulations instead of the chromosome length. But, the behavior in this PR is to clip the recombination map to the size of the contig. I can modify the PR to keep the current behavior (when the contig is the entire chromosome and the recombination map is longer) but this seems like an odd edge case to document. Especially if the plan is to deprecate genetic maps that don't match the assembly. So I've kept the new "clip-to-contig" behavior.

There's no effect on any of the tests, aside from the occasional stochastic failure of `test_slim_engine.py::TestRecombinationMap::test_chr1` that uses "HapMapII_GRCh37" and explicitly checks that simulated tree sequence breakpoints fall within the recombination map length. Updating this test to use "HapMapII_GRCh38" eliminates the stochastic failure.

File: stdpopsim/genomes.py
Comment (grahamgower) [inline stdpopsim/genomes.py:396]: Agreed. I vote to deprecate the length multiplier. It might have been useful during initial testing and development, but it doesn't compose well with any of the newer Contig features.

File: stdpopsim/genomes.py
Comment (grahamgower) [inline stdpopsim/genomes.py:414]: Is there a reason not to just explicitly raise an error in this case?

File: stdpopsim/genomes.py
Comment (grahamgower) [inline stdpopsim/genomes.py:473]: I think it would be better to store the original coordinates as Contig fields and produce the `origin` property from them, instead of the storing and parsing the origin string.

File: stdpopsim/utils.py
Comment (grahamgower) [inline stdpopsim/utils.py:267]:
```suggestion
    for (il, ir) in intervals:
```

File: tests/test_masking.py
Comment (grahamgower) [inline tests/test_masking.py:16]:
```suggestion
    @pytest.usefixtures("tmp_path")
    def test_load_intervals(self, tmp_path):
```
I know this isn't required, but unless one is very familiar with pytest then reading tests with "magic" arguments is very confusing.

**Review by grahamgower — CHANGES_REQUESTED** [review]
This is great, thanks @nspope!

I'm luke warm on all the clipping though. Can these just raise errors instead? Are there any reasonable use cases for mismatched masks, recombination maps, and genome coordinates?

File: stdpopsim/genomes.py
Comment (nspope) [reply stdpopsim/genomes.py:414]: Hmm, say there's an available chromosome-wide mask and annotation, and the user wants to use these whilst simulating some small region. One option is that the user clips the mask/annotation to the region themselves, and stdpopsim raises errors if these don't all fall within the region. Another option is that stdpopsim does the clipping and warns if something extreme happens, like all mask/annotation intervals falling outside of the region (this is the current strategy in the PR). I could go either way -- not sure what the right trade-off is between ease-of-use and preventing miss-specification.

Another possibility is to go the latter route (stdpopsim does the clipping-to-region) but first check that the mask/annotation matches the chromosome boundaries and raise an error if not.

File: stdpopsim/genomes.py
Comment (nspope) [reply stdpopsim/genomes.py:396]: I agree. So, leave the length multiplier feature in there (with a deprecation warning?) but don't try to accommodate it when adding new arguments/features.

**Comment by nspope** [comment]
Thanks @grahamgower -- with regards to clipping, say there's an available chromosome-wide mask and annotation, and the user wants to use these whilst simulating some small region. One option is that the user clips the mask/annotation to the region themselves, and stdpopsim raises errors if any of the provided intervals fall outside the region. Another option is that stdpopsim does the clipping and warns if something extreme happens, like all mask/annotation intervals falling outside of the region (this is the current strategy in the PR). I could go either way -- not sure what is the right trade-off between ease-of-use and preventing miss-specification.

Another possibility is to go the latter route (stdpopsim does the clipping-to-region) but first check that the mask/annotation matches the chromosome boundaries and raise an error if not.

File: stdpopsim/genomes.py
Comment (grahamgower) [reply stdpopsim/genomes.py:396]: Yeah, exactly. Don't worry about deprecating it in this pr though.

File: stdpopsim/genomes.py
Comment (grahamgower) [reply stdpopsim/genomes.py:414]: Ok, I get it. In that case I think you've made the right decision.

**Comment by nspope** [comment]
I think the last round of commits should address your comments, @grahamgower.

@andrewkern and/or @petrelharp, could you also take a look? It'd be good to get another opinion with regards to (1) keeping coordinate system of parent chromosome for adding DFEs; (2) clipping masks/maps/dfes to the contig boundaries.

**Review by grahamgower — APPROVED** [review]
LGTM!

**Review by andrewkern — COMMENTED** [review]
just some minor changes here but looks good to me

**Comment by andrewkern** [comment]
we'll also want to add some docs for this including a use case example

File: stdpopsim/cli.py
Comment (nspope) [reply stdpopsim/cli.py:624]: yeah, they can. I'll remove that bit, everything is on the same coordinate system so I don't think we need to confuse things by talking about the coordinate system.

File: stdpopsim/genomes.py
Comment (nspope) [reply stdpopsim/genomes.py:256]: It's all converted under the hood to numpy arrays now -- see this test (which was skipped previous to this PR):

https://github.com/popsim-consortium/stdpopsim/blob/2e042a7a04d97896bfa3f076df8f1963efdf3f5b/tests/test_genomes.py#L113

and _check_intervals_validity will complain if it's not the right shape. 

Or do you think we want to make the input format specific?

File: stdpopsim/genetic_maps.py
Comment (nspope) [reply stdpopsim/genetic_maps.py:126]: Fair enough-- but I do think we want to deprecate/discourage using old genetic maps so let's keep a TODO there?

**Comment by nspope** [comment]
in addition to use case, we'll need to make it clear in the docs that simulating a region with selection is not the same as simulating a chromosome with selection and cropping to the region.

**Comment by nspope** [comment]
@andrewkern do you mind taking a look at the additions to the tutorial?

Otherwise I think this is good to go!

**Review by andrewkern — APPROVED** [review]

**Comment by andrewkern** [comment]
looks good to me @nspope! can you rebase and squash these changes before we merge?

**Comment by petrelharp** [comment]
> rebase and squash these changes before we merge?

"Squash and merge" will do this for you.

**Pull request merged**
